File: app/src/main/java/com/aero51/moviedatabase/ui/adapter/PeopleSearchPagedListAdapter.java
package com.aero51.moviedatabase.ui.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.paging.PagedListAdapter;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.aero51.moviedatabase.R;
import com.aero51.moviedatabase.repository.model.tmdb.credits.ActorSearchResponse;
import com.squareup.picasso.Picasso;

import de.hdodenhof.circleimageview.CircleImageView;

import static com.aero51.moviedatabase.utils.Constants.BASE_IMAGE_URL;
import static com.aero51.moviedatabase.utils.Constants.PROFILE_SIZE_W185;


public class PeopleSearchPagedListAdapter extends PagedListAdapter<ActorSearchResponse.ActorSearch, PeopleSearchPagedListAdapter.ViewHolder> {
    private CastAdapter.ItemClickListener mClickListener;


    public PeopleSearchPagedListAdapter() {
        super(DIFF_CALLBACK);
    }


    @NonNull
    @Override
    public PeopleSearchPagedListAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        View view = layoutInflater.inflate(R.layout.actor_search_item, parent, false);
        return new PeopleSearchPagedListAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull PeopleSearchPagedListAdapter.ViewHolder holder, int position) {
        ActorSearchResponse.ActorSearch person = getItem(position);
        String imageUrl = BASE_IMAGE_URL + PROFILE_SIZE_W185 + person.getProfile_path();
        Picasso.get().load(imageUrl).into(holder.castProfileImageView);
        holder.textViewRealName.setText(person.getName());
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        CircleImageView castProfileImageView;
        TextView textViewRealName;

        ViewHolder(View itemView) {
            super(itemView);
            castProfileImageView = itemView.findViewById(R.id.cast_profile_image_view);
            textViewRealName = itemView.findViewById(R.id.text_view_real_name);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            Integer adapter_position = getBindingAdapterPosition();
            if (mClickListener != null)
                mClickListener.onItemClick(view, getItem(adapter_position).getId(), adapter_position);
        }
    }

    // allows clicks events to be caught
    public void setClickListener(CastAdapter.ItemClickListener itemClickListener) {
        this.mClickListener = itemClickListener;
    }

    // parent activity will implement this method to respond to click events
    public interface ItemClickListener {
        void onItemClick(View view, Integer actorId, int position);
    }


    private static DiffUtil.ItemCallback<ActorSearchResponse.ActorSearch> DIFF_CALLBACK =
            new DiffUtil.ItemCallback<ActorSearchResponse.ActorSearch>() {
                @Override
                public boolean areItemsTheSame(ActorSearchResponse.ActorSearch oldItem, ActorSearchResponse.ActorSearch newItem) {
                    return oldItem.getId().equals(newItem.getId());
                }

                @Override
                public boolean areContentsTheSame(ActorSearchResponse.ActorSearch oldItem, ActorSearchResponse.ActorSearch newItem) {
                    return oldItem.getName().equals(newItem.getName());
                }
            };
}

File: app/src/main/java/com/aero51/moviedatabase/ui/adapter/PeopleSearchPagedListAdapter.kt
package com.aero51.moviedatabase.ui.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.aero51.moviedatabase.R
import com.aero51.moviedatabase.repository.model.tmdb.credits.ActorSearchResponse.ActorSearch
import com.aero51.moviedatabase.utils.ActorClickListener
import com.aero51.moviedatabase.utils.Constants.BASE_IMAGE_URL
import com.aero51.moviedatabase.utils.Constants.PROFILE_SIZE_W185
import com.aero51.moviedatabase.utils.ObjectClickListener
import com.squareup.picasso.Picasso
import de.hdodenhof.circleimageview.CircleImageView

class PeopleSearchPagedListAdapter(private val itemClickListener: ActorClickListener) : PagedListAdapter<ActorSearch, PeopleSearchPagedListAdapter.ViewHolder>(DIFF_CALLBACK) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        val view = layoutInflater.inflate(R.layout.actor_search_item, parent, false)
        return ViewHolder(view,itemClickListener)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val person = getItem(position)
        val imageUrl: String = BASE_IMAGE_URL + PROFILE_SIZE_W185 + person!!.profile_path
        Picasso.get().load(imageUrl).into(holder.castProfileImageView)
        holder.textViewRealName.text = person.name
    }

    inner class ViewHolder internal constructor(itemView: View, itemClickListener: ActorClickListener?) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        var castProfileImageView: CircleImageView
        var textViewRealName: TextView
        private val itemClickListener: ActorClickListener?
        init {
            castProfileImageView = itemView.findViewById(R.id.actor_profile_image_view)
            textViewRealName = itemView.findViewById(R.id.text_view_actor_search_name)
            this.itemClickListener = itemClickListener
            itemView.setOnClickListener(this)
        }
        override fun onClick(view: View) {
            val adapter_position = bindingAdapterPosition
            if (itemClickListener != null) {
                getItem(adapter_position)?.let { itemClickListener.onActorItemClick(it.id, adapter_position) }
            } // call the onClick in the OnItemClickListener
        }


    }


    companion object {
        private val DIFF_CALLBACK: DiffUtil.ItemCallback<ActorSearch> = object : DiffUtil.ItemCallback<ActorSearch>() {
            override fun areItemsTheSame(oldItem: ActorSearch, newItem: ActorSearch): Boolean {
                return oldItem.id == newItem.id
            }

            override fun areContentsTheSame(oldItem: ActorSearch, newItem: ActorSearch): Boolean {
                return oldItem.name == newItem.name
            }
        }
    }
}